src/boardState.ts

```diff
@@ -0,0 +1,239 @@
+import {ShipPlacement} from './shipPlacement';
+import {ShipPossibilities} from './shipPossibilities';
+import * as Globals from './constants';
+import {Position} from './position';
+import {Shot} from './shot';
+import {randBetween} from './random'
+import {ShotImpliedPlacements} from './shotImpliedPlacements'
+
+export class BoardState{
+    public ships: Array<ShipPlacement> = [];
+    public shots: Array<Shot> = [];
+
+    public constructor(ships: Array<ShipPlacement>, shots: Array<Shot> = []){
+        this.ships = ships;
+        this.shots = shots;
+    }
+
+    static tryGetRandomSet(shipLengths: Array<number>, shots:Array<Shot> = [], hitShots: Array<Shot> = [], pathToFollow: number[] = [], exhaustive:boolean = false):{state: BoardState, found: boolean, nextPath: number[]}
+    {
+        let possibleShips: Array<ShipPlacement> = [];
+        let validShipPositions: Array<ShipPossibilities> = [];
+
+        let numberRemaining: {[length: number]: number} = [];
+
+        let shotImpliedPlacementsList: Array<ShotImpliedPlacements> = [];
+
+        let currentPathIndex = 0; //how many guesses deep it is
+        let currentPath: Array<number> = []; //the path of guesses it is currently following
+
+        for(var i=0; i<shipLengths.length; i++){
+            validShipPositions.push(new ShipPossibilities(shipLengths[i]));
+            validShipPositions[i].removeFromShots(shots);
+
+            if(!numberRemaining[shipLengths[i]]){
+                numberRemaining[shipLengths[i]] = 1;
+            }
+            else{
+                numberRemaining[shipLengths[i]]++;
+            }
+        }
+
+        for(var i=0; i<hitShots.length; i++){
+            shotImpliedPlacementsList.push(new ShotImpliedPlacements(hitShots[i]));
+            shotImpliedPlacementsList[i].removeInvalidPositions(validShipPositions);
+        }
+
+        shotImpliedPlacementsList.sort((a,b,)=>{return a.allValidPlacements.length - b.allValidPlacements.length});
+
+        //sort by #
```

Comment:
If you're putting in comments to explain what the intent of part of a method is, that's a really strong reason to start extracting those sections into named methods.